src/realm/sync/client.hpp

```diff
@@ -40,13 +41,6 @@ class Client {
     Client(Client&&) noexcept;
     ~Client() noexcept;
 
-    /// Run the internal event-loop of the client. At most one thread may
-    /// execute run() at any given time. The call will not return until somebody
-    /// calls stop().
-    void run() noexcept;
```

Comment:
I was thinking that you'd add a method like `Future<void> Session::shutdown()` that works like [`void Session::detach()`](https://github.com/realm/realm-core/blob/64be33187a3aadefd39fb83c7dc46e125f98c2b3/src/realm/sync/client.hpp#L397) but returns a future when the `SessionWrapper` is finalized. Maybe we can even just make `Session::detach()` return a `Future<void>` that is filled when the session is finalized? Whatever we do, I think we need to wait until we're [here](https://github.com/realm/realm-core/blob/64be33187a3aadefd39fb83c7dc46e125f98c2b3/src/realm/sync/client.cpp#L1749) before we declare the Session properly torn down.

Reply:
There are many ways to go about this. I followed the `shutdown_and_wait` API and adapted it as I thought it makes the most sense. We do wait until `finalize()` completes as that is called from the event loop and the callback we post runs after that. If we use a future, we need to fulfill it from the event loop to make sure the events happen sequentially.